pkg/networkserver/grpc_deviceregistry.go

```diff
@@ -275,392 +269,1969 @@ func (ns *NetworkServer) Get(ctx context.Context, req *ttnpb.GetEndDeviceRequest
+		if !m[left].FieldIsZero(left) {
+			return true, ""
+		}
+		return isValid(m)
+	}, append([]string{left}, right...)...)
+}
+
+// ValidateIfNotZeroThenFunc ensures !FieldIsZero(left) -> f(map r -> *ttnpb.EndDevice), for each r in right.
+func (st *setDeviceState) ValidateIfNotZeroThenFunc(isValid func(map[string]*ttnpb.EndDevice) (bool, string), left string, right ...string) error {
+	if st.HasSetField(left) {
+		if st.Device.FieldIsZero(left) {
+			return nil
+		}
+		return st.ValidateFields(isValid, right...)
+	}
+	if !st.HasSetField(right...) {
+		return nil
+	}
+	return st.ValidateFields(func(m map[string]*ttnpb.EndDevice) (bool, string) {
+		if m[left].FieldIsZero(left) {
+			return true, ""
+		}
+		return isValid(m)
+	}, append([]string{left}, right...)...)
+}
+
+// SetFunc is the function meant to be passed to SetByID.
+func (st *setDeviceState) SetFunc(f func(context.Context, *ttnpb.EndDevice) error) func(context.Context, *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
+	return func(ctx context.Context, stored *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
+		for p, shouldBeZero := range st.zeroPaths {
+			if stored.FieldIsZero(p) != shouldBeZero {
+				return nil, nil, newInvalidFieldValueError(p)
+			}
+		}
+		for _, g := range st.onGet {
+			if err := g(stored); err != nil {
+				return nil, nil, err
+			}
+		}
+		if err := f(ctx, stored); err != nil {
+			return nil, nil, err
+		}
+		return st.Device, st.SetFields(), nil
+	}
+}
+
+func newSetDeviceState(dev *ttnpb.EndDevice, paths ...string) *setDeviceState {
+	return &setDeviceState{
+		Device: dev,
+		paths:  paths,
+
+		pathsCache:     make(map[string]bool),
+		extraSetsCache: make(map[string]bool),
+		extraGetsCache: make(map[string]bool),
+
+		zeroPaths: make(map[string]bool),
+	}
+}
+
+func setKeyIsZero(m map[string]*ttnpb.EndDevice, get func(*ttnpb.EndDevice) *ttnpb.KeyEnvelope, path string) bool {
+	if dev, ok := m[path+".key"]; ok {
+		if ke := get(dev); !ke.Key.IsZero() {
+			return false
+		}
+	}
+	if dev, ok := m[path+".encrypted_key"]; ok {
+		if ke := get(dev); len(ke.EncryptedKey) != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func setKeyEqual(m map[string]*ttnpb.EndDevice, getA, getB func(*ttnpb.EndDevice) *ttnpb.KeyEnvelope, pathA, pathB string) bool {
+	if a, b := getA(m[pathA+".key"]).GetKey(), getB(m[pathB+".key"]).GetKey(); a == nil && b != nil || a != nil && b == nil || a != nil && b != nil && !a.Equal(*b) {
+		return false
+	}
+	if a, b := getA(m[pathA+".encrypted_key"]).GetEncryptedKey(), getB(m[pathB+".encrypted_key"]).GetEncryptedKey(); !bytes.Equal(a, b) {
+		return false
+	}
+	if a, b := getA(m[pathA+".kek_label"]).GetKEKLabel(), getB(m[pathB+".kek_label"]).GetKEKLabel(); a != b {
+		return false
+	}
+	return true
+}
+
+// Set implements NsEndDeviceRegistryServer.
+func (ns *NetworkServer) Set(ctx context.Context, req *ttnpb.SetEndDeviceRequest) (*ttnpb.EndDevice, error) {
```

Comment:
This one is now 1570 lines long without any comments for logical sections. Also, there are some long lists of field paths; how do we maintain those? Can't those be put in a `var` with a godoc with instructions on how to keep that list up to date (like what the rule of for having that path there)?

Reply:
https://github.com/TheThingsNetwork/lorawan-stack/pull/3680/commits/c1d8e5ea8b7f8e6a5663e6d0c7295076ab329e8e what do you think about this?

Reply:
Yes, that's much better